src/dcd/server/autocomplete/ufcs.d

```diff
@@ -0,0 +1,253 @@
+
+    assert(beforeDotSymbol);
+
+    if (beforeDotSymbol.name is getBuiltinTypeName(tok!"void")
+        || (beforeDotSymbol.type !is null
+            && beforeDotSymbol.type.name is getBuiltinTypeName(tok!"void")))
+    {
+
+        return null; // no UFCS for void
+    }
+
+    Scope* currentScope = completionScope.getScopeByCursor(cursorPosition);
+    assert(currentScope);
+    HashSet!size_t visited;
+    // local imports only
+    FilteredAppender!(a => a.isCallableWithArg(beforeDotSymbol), DSymbol*[]) app;
+    while (currentScope !is null && currentScope.parent !is null)
+    {
+        auto localImports = currentScope.symbols.filter!(a => a.kind == CompletionKind.importSymbol);
+        foreach (sym; localImports)
+        {
+            if (sym.type is null)
+                continue;
+            if (sym.qualifier == SymbolQualifier.selectiveImport)
+                app.put(sym.type);
+            else
+                sym.type.getParts(internString(null), app, visited);
+        }
+
+        currentScope = currentScope.parent;
+    }
+    // global symbols and global imports
+    assert(currentScope !is null);
+    assert(currentScope.parent is null);
+    foreach (sym; currentScope.symbols)
+    {
+        if (sym.kind != CompletionKind.importSymbol)
+            app.put(sym);
+        else if (sym.type !is null)
+        {
+            if (sym.qualifier == SymbolQualifier.selectiveImport)
+                app.put(sym.type);
+            else
+                sym.type.getParts(internString(null), app, visited);
+        }
+    }
+    return app.data;
+}
+
+/**
+   Params:
+   symbol = the symbol to check
+   arg0 = the argument
+   Returns:
+   true if if $(D symbol) is callable with $(D arg0) as it's first argument
+   false otherwise
+*/
+bool isCallableWithArg(const(DSymbol)* symbol, const(DSymbol)* arg0)
+{
+    // FIXME: do signature type checking?
+    // 	a lot is to be done in dsymbol for type checking to work.
+    //  for instance, define an isSbtype function for where it is applicable
+    // 	ex: interfaces, subclasses, builtintypes ...
+
+    // FIXME: instruct dsymbol to always save paramater symbols
+    // 	 and check these instead of checking callTip
+
+    static bool checkCallTip(string callTip)
+    {
+        assert(callTip.length);
+        if (callTip.endsWith("()"))
+            return false; // takes no arguments
+        else if (callTip.endsWith("(...)"))
+            return true;
+        else
+            return true; // FIXME: assume yes?
+    }
+
+    assert(symbol);
+    assert(arg0);
+
+    switch (symbol.kind)
+    {
+    case CompletionKind.dummy:
+        if (symbol.qualifier == SymbolQualifier.func)
+            return checkCallTip(symbol.callTip);
+        break;
+    case CompletionKind.importSymbol:
+        if (symbol.type is null)
+            break;
+        if (symbol.qualifier == SymbolQualifier.selectiveImport)
+            return isCallableWithArg(symbol.type, arg0);
```

Comment:
Instead use `symbol.type.isCallableWithArg(arg0)`

Reply:
I see you do like ufcs :)

Reply:
Hey, and this MR makes it even more motivating to use it.